Use TypeScript

felixfbecker opened the pull request:
This is a PR to change the implementation from transpiled ES6 code to transpiled TypeScript code.

The biggest advantage is type safety for properties, call signatures, return types and object interfaces, for the code in this repo, implementations and consumers.
Code is transpiled to ES5 code just like before, so for JavaScript consumers nothing changes.
TypeScript users can now use the package with TypeScript and get generated definition files (.d.ts) for free.

As TypeScript is a superset of JavaScript, the actual changes are pretty small. All that's added is a few type declarations and visibilities for `protected` methods that are intended to be overridden by implementations.

TODO:
- [x] Adapt code
- [x] Adapt tests
- [x] Add TSLint
- [x] Remove Babel
- [x] Configure webpack
- [x] Add generated documentation with tsdoc
- [x] Update README

yurishkuro commented on src/tracer.ts:
```diff
@@ -1,8 +1,48 @@
+    startTime?: number;
+}
+
+export type Format = typeof Constants.FORMAT_BINARY | typeof Constants.FORMAT_TEXT_MAP | typeof Constants.FORMAT_HTTP_HEADERS;
```
will this restrict the format argument to just these three? It is meant to be open-ended.

felixfbecker replied:
Oh, wasn't aware of that. This would indeed restrict it to these three. An implementation can of course overload the signature.

yurishkuro commented on src/tracer.ts:
```diff
@@ -42,28 +82,10 @@ export default class Tracer {
      * @return {Span} - a new Span object.
      */
-    startSpan(name, fields) {
+    startSpan(name: string, fields: SpanFields = {}): Span {
```
I'd prefer `s/fields/options/`, it seems to be a common pattern in JS libs and it is more accurate than "fields". Similarly, `s/SpanFields/StartSpanOptions/`

felixfbecker replied:
Totally agree, since only a defined set of properties are accepted.

yurishkuro commented on src/tracer.ts:
```diff
@@ -130,7 +151,10 @@ export default class Tracer {
      */
-    inject(spanContext, format, carrier) {
+    inject(spanContext: SpanContext | Span, format: typeof Constants.FORMAT_TEXT_MAP, carrier: { [key: string]: string }): void;
+    inject(spanContext: SpanContext | Span, format: typeof Constants.FORMAT_BINARY, carrier: { buffer: number[] }): void;
```
is `number[]` the idiomatic representation of a byte buffer?

felixfbecker replied:
I wasn't sure because the documentation doesn't specify this. It just says an "Array-like" that carries binary data. I interpreted that as a byte array, eg https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array or NodeJS Buffer (which would be compatible with `number[]` / `Array<number>`)

yurishkuro commented on src/tracer.ts:
```diff
@@ -130,7 +151,10 @@ export default class Tracer {
+    inject(spanContext: SpanContext | Span, format: typeof Constants.FORMAT_TEXT_MAP, carrier: { [key: string]: string }): void;
+    inject(spanContext: SpanContext | Span, format: typeof Constants.FORMAT_BINARY, carrier: { buffer: number[] }): void;
+    inject(spanContext: SpanContext | Span, format: typeof Constants.FORMAT_HTTP_HEADERS, carrier: { [key: string]: string }): void;
+    inject(spanContext: SpanContext | Span, format: Format, carrier: any): void {
```
Doesn't typescript support generics? Java API uses generics to enforce type parity between format and carrier without having to define overloaded methods. In practice many tracer implementations do a lookup in the internal map for injectors/extractors corresponding to the appropriate format, so forcing them to implement 4 methods is not necessary.

felixfbecker replied:
TypeScript does support both generic classes and methods. What I wanted to accomplish here was to enforce that if you eg pass `FORMAT_BINARY`, the `carrier` needs to be an object with a `buffer` property.
Overloads in TypeScript are not matched automatically, you just define the overloaded signatures, and then you must define one "generic" implementation that usually does runtime type checks. Implementations do not need to implement the 4 overloads.

yurishkuro commented:
I kinda like this (admittedly, I'm biased by my preference for strongly typed over dynamic languages). And the transpiled JS is not horrible at all, it's actually even better than Babel version as it assigns to prototypes directly instead of using a table-driven initialization.

felixfbecker commented:
Glad you like the general direction. I think the source is still easy to understand and contribute to for native JavaScript developers, since it really is just JavaScript + types. Emitting readable JS is one of TS' defined language goals, so you can count on that ;)

tedsuo commented:
I'm all for moving to TypeScript. Please make sure not to change the API implicitly while lifting up on to it.

felixfbecker commented:
I addressed the feedback, added typedoc documentation, configured webpack, removed Babel and lib-debug

bhs commented:
@felixfbecker FYI, I have been time-bankrupt but finally took a look at this.

TL;DR "thank you". I should not have much weight in this discussion since (a) I am just a journeyman-level javascript programmer and (b) the change here is not about OT semantics, but rather js ecosystem best practices. That said, the .ts and .js for the important files is/are all readable.

bhs commented:
Oh, one request, though: please add something to the README about how both Typescript and non-Typescript javascript programmers should interact with this code (esp the non-Typescript people who might be scared off without some docs pointing them towards `lib/tracer.js` or whatever).

felixfbecker commented:
I think this is ready for final review. I adapted the tests, updated the readme and removed the compiled JS from git.

I feel like after this is merged it would be a very good opportunity to release v1.

felixfbecker commented:
Also, in case someone is interested, https://github.com/sourcegraph/javascript-typescript-langserver is completely instrumented using this branch.
It is used together with async/await and Observables and also both injects and extracts contexts into/from JSON RPC messages.

felixfbecker commented:
Important step needed for the documentation: I've changed the docs generator to typedoc and added a deployment to travis to https://opentracing-javascript.surge.sh. Someone needs to publish this once, then set the secret SURGE_LOGIN and SURGE_TOKEN env vars in the travis settings.

felixfbecker commented:
ping @yurishkuro @tedsuo @bhs

bhs commented on README.md:
```diff
@@ -77,71 +77,65 @@ http.request(opts, function (res) {
 * `dist/opentracing-browser.min.js` - minified, no runtime checks
 * `dist/opentracing-browser.js` - debug version with runtime checks
 
+### Usage with TypeScript
```
thank you

felixfbecker commented:
In theory it's a feature. But I would actually suggest to bump major with this for a couple of reasons:
- I removed runtime type checks
- The lib-debug.js entry point is now missing, so if you imported from that, it won't work anymore
- There are some slight differences in module exports. Babel (incorrectly) exported the whole exports object as the default export, which allows to import with `import opentracing from 'opentracing'` instead of `import * as opentracing from 'opentracing'`. I saw this in the LightStep codebase (the runtime tracer still works though)
- A library is not under semver below 1.0 and considered experimental. I am sure we all would feel more comfortable with a 1.0 release in production, and I feel like this is a good starting point.

Also make sure to configure surge before merging, or the build will fail.

bcronin commented:
@felixfbecker , just wanted to add that this is awesome work

yurishkuro commented on src/test/api_compatibility.ts:
```diff
@@ -0,0 +1,93 @@
+
```
This is a comment about span_context.ts (which I can't comment on directly).
* `default class` was replaced in other files
* can it be just an interface?

felixfbecker replied:
If it's an interface, then it's not an noop implementation anymore.

yurishkuro commented on src/test/api_compatibility.ts:
```diff
@@ -0,0 +1,93 @@
+            describe('startSpan', () => {
+                it('should handle Spans and SpanContexts', () => {
+                    expect(() => { tracer.startSpan('child', { childOf : span }); }).to.not.throw(Error);
+                    expect(() => { tracer.startSpan('child', { childOf : span.context() }); }).to.not.throw(Error);
```
add followsFrom?

felixfbecker replied:
I only converted this file to TS. Tbh I don't even know what a test is useful for that only tests that a function doesn't throw an exception, it doesn't verify any behaviour at all

yurishkuro replied:
It can be useful for pure JS implementations. It checks that a given implementation satisfies the API, and in some cases checks the behavior, like chaining of methods, and baggage.

yurishkuro commented on src/test/api_compatibility.ts:
```diff
@@ -0,0 +1,93 @@
+                    expect(() => { tracer.extract(FORMAT_TEXT_MAP, textCarrier); }).to.not.throw(Error);
+                    expect(() => { tracer.extract(FORMAT_BINARY, binCarrier); }).to.not.throw(Error);
+                    expect(() => { tracer.extract(FORMAT_BINARY, {}); }).to.not.throw(Error);
+                    expect(() => { tracer.extract(FORMAT_BINARY, { buffer : null }); }).to.not.throw(Error);
```
not clear why empty and null binary carriers are value combinations

felixfbecker replied:
Not clear to me either

yurishkuro commented on src/tracer.ts:
```diff
@@ -0,0 +1,188 @@
+     */
+    startSpan(name: string, options: SpanOptions = {}): Span {
+
+        // Convert fields.childOf to fields.references as needed.
```
`s/field./options./`

yurishkuro commented on src/test/opentracing_api.ts:
```diff
@@ -0,0 +1,79 @@
+import { expect } from 'chai';
+import * as opentracing from '../index';
+
+export function opentracingAPITests(): void {
```
Q: assuming we keep ot.Tracer as a concrete class, why can't these checks be in the api-compatibility file?

felixfbecker replied:
Tbh I don't see any value in these tests. They just verify that exports and methods exist. This is nothing the compiler can't warn you about.

yurishkuro commented on src/test/opentracing_api.ts:
```diff
@@ -0,0 +1,79 @@
+        describe('Reference', () => {
+            it('should be a class', () => {
+                const ref = new opentracing.Reference(opentracing.REFERENCE_CHILD_OF, span.context());
+                expect(ref).to.be.an('object');
```
can this ever be not an object, given that it's created with `new`?

felixfbecker replied:
No. This test is pretty useless.

yurishkuro commented on src/tracer.ts:
```diff
@@ -0,0 +1,188 @@
+     * overrides the first argument to startSpan(). Provided for
+     * backwards-compatibility.
+     */
+    operationName?: string;
```
imo we should drop it, especially if we are going for v1.0

bhs replied:
+1

felixfbecker commented on README.md:
```diff
@@ -42,32 +42,32 @@ Spans:
     });
-    res.on('data', function (chunk) {
+    res.on('data', chunk => {
         span.log({'event': 'data_received', 'data': chunk});
```
sure, but the example was there before. I only converted the example to use latest JS language features.

felixfbecker commented on README.md:
```diff
@@ -77,71 +77,65 @@ http.request(opts, function (res) {
 
-Implementations can subclass `opentracing.Trace`, `opentracing.Span`, and the other API classes to build a OpenTracing tracer.
-
-Due to the dynamic nature of JavaScript, implementations can simply implement classes with the same signatures as the OpenTracing classes and use these directly as well (there's no need to subclass).
```
It depends. If you use TypeScript, and type your method to take an opentracing `Span`, it will work as long as the implementation has the same methods and properties. For JS, it doesn't matter of course. But I didn't get why this needs to be highlighted here, that is simply how JS works and should be obvious to any JS dev. I didn't want to confuse people by explaining the exact circumstances, but rather encourage implementations to subclass the opentracing implementations. What do you mean with going through a proxy class?

felixfbecker commented on scripts/docker_test.sh:
```diff
@@ -2,4 +2,4 @@
 
 docker run -t --rm --name test-runner-opentracing-javascript \
```
I cannot answer that @yurishkuro, I didn't write this. I just adapted it to not use the Makefile

felixfbecker commented on src/binary_carrier.ts:
```diff
@@ -5,7 +5,10 @@
+
+    public buffer?: ArrayLike<number>;
+
+    constructor(public binaryData?: ArrayLike<number>) {
```
Agree, but again, I just converted it to TypeScript. It is optional because there are test cases that call this without passing any parameter.

felixfbecker commented on src/reference.ts:
```diff
@@ -1,6 +1,9 @@
 import Span from './span';
 import SpanContext from './span_context';
 
+export type ReferenceType = typeof REFERENCE_CHILD_OF | typeof REFERENCE_FOLLOWS_FROM;
```
Should I let `type()` return `string` instead?

yurishkuro replied:
I would just use strings, similar to how Format is handled.

In other languages we sometimes were able to define a type that was externally extensible, e.g. in Go. But in Java it's just a string, even though the [Format](https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/propagation/Format.java#L36) is strongly typed.

felixfbecker commented on src/span.ts:
```diff
@@ -1,12 +1,14 @@
  * may have zero or more child Spans, which in turn may have children.
  */
-export default class Span {
+export class Span {
```
I don't really understand your concern. These classes _are_ already the noop implementation. What coupling?
For example, if you look at how I used them here: https://sourcegraph.com/github.com/sourcegraph/javascript-typescript-langserver/-/blob/src/project-manager.ts#L264:1-265:1
I declare each method to take a `Span` that defaults to a noop `Span` instance, and I can pass e.g. a LightStep `Span` just fine because it is compatible.

I totally agree that having all these underscore-prefixed methods is kind of useless, they serve no purpose besides little things like not having to convert `Span | SpanContext` to `SpanContext`. Implementations could just override the public methods. That would be a breaking change though.

What we could do is do a complete restructure, have all these now-classes be interfaces (those will not be translated to JS in any way), and then add classes like `NoopSpan`, `NoopTracer` etc. You would use them like this then:
```ts
import { Span, NoopSpan } from 'opentracing'
function foo(span: Span = new NoopSpan()) {

}

yurishkuro replied:
By coupling I mean the workarounds with initialization for Noop due to circular dependencies. If `opentracing.Tracer` itself is meant to be a concrete implementation, then it has to be a noop implementation, so there's no reason to have Noop as something separate, it just complicates the library.

I am not too concerned with backwards compatibility given v1.0 label (the external API is still the same). But we can keep some of the shared functionality like normalizing references in StartSpan options. I'd just prefer to see a lot fewer underscore methods, especially when the delegation is completly pass through, as pretty much in all Span methods.

felixfbecker replied:
Well, as a TS user, I can only use this as soon as the implementation of my choice adopts it. So I wouldn't say it doesn't matter, but I would rather have a cleaner API.

But the outer API _would_ change. You would have to export the noop implementation under a different name if e.g. `Tracer` is now just an interface (look at my example above).

Noop isn't something separate right now really, all the classes are documented of implementing noop behavior, and I already used them that way

felixfbecker commented:
@yurishkuro I addressed your feedback. While I agree that we could change everything to interfaces, I am not clear about how exactly you would want it to look like.
 - we could have a folder `interfaces` and a folder `noop`, with all noop implementations.
 - what happens to the not-noop implementations, like `Reference`?
 - shall we drop all underscore-prefixed methods then?
 - what happens to normalisation that happened there? Are implementations expected to do those themselves? What do others do here (e.g. opentracing-java)?

This would change the change from a _slight_ breaking change to a major breaking change - which means both more effort I would have to put in and longer adoption time for implementations. The PR is already pretty big by now, so how about we release this as 0.14 and switch to interfaces for 1.0?

yurishkuro commented on README.md:
```diff
@@ -77,71 +77,65 @@ http.request(opts, function (res) {
-* `make build` creates the compiled, distributable code
-* `make test` runs the tests
+* `npm run build` creates the compiled, distributable JavaScript code in `./lib`
+* `npm run watch` incrementally sompiles on file changes
```
ping

felixfbecker replied:
Sorry, fixed.

yurishkuro commented:
I'm ok to merge this and consider other changes separately.

bhs commented:
@felixfbecker last thing: please actually update the minor version in the package.json and also update the CHANGELOG. After that I'm :thumbsup: to merge this. Thanks again, nearly there...

felixfbecker commented:
Okay. I would usually do this on the master branch with `npm version minor`, because it commits the version increase in package.json + git tags the version automatically.

Also, as a reminder, for the documentation someone needs to:
- check this branch out
-`npm run typedoc`
-`npm install -g surge`
- cd into the typedoc folder, run `surge`
- publish to the domain `opentracing-javascript.surge.sh`
- get an access token with `surge token`
- add the env vars `SURGE_LOGIN` (your email) and `SURGE_TOKEN` (the generated token) to the secret env vars on Travis

Future updates will be deployed automatically.

bhs commented:
@felixfbecker ok, sounds like a plan re `npm version minor`. I will try to do that and the remaining steps you suggested when I actually have a block to do it in (i.e., not now). Many thanks again, this PR is a substantial improvement IMO.

bhs commented:
@felixfbecker you had conflicts against master, but github let me rebase your branch and push to the `felixfbecker/opentracing-javascript` repo (which I have done). I'll try merging and going through the steps you outlined above.

felixfbecker commented:
Watch out, the changes on master were updates to docblocks and a removal of a dependency - it's very likely that these are reverted by this PR because the diff recreates a lot of files

bhs commented:
@felixfbecker well, I probably effed that up nicely then – I really only had about 30 mins to do all of this in the midst of a bunch of other stuff. For better or worse I was successful at everything in https://github.com/opentracing/opentracing-javascript/pull/68#issuecomment-296284685 and also the minor version bump. :-/

See http://opentracing-javascript.surge.sh/ and let me know if you need help fixing anything that's broken.

felixfbecker commented:
master build is broken because the surge deploy is broken:

> Please add SURGE_TOKEN in Travis settings (get your token with 'surge token')
> failed to deploy

I saw you added encrypted values to travis.yml, I don't know how these work, for secret env vars I always just set them in the settings: https://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-Repository-Settings

I will look at which changes were overridden and reapply them tomorrow, it's late night here already.

bhs commented:
@felixfbecker 

I fixed the master build – was a stupid yaml formatting mistake.

bhs commented:
(It's hard to test travis stuff without pushing code – not a great state of affairs)

felixfbecker commented:
Please next time don't force push the branch, I depended on a specific commit hash from that branch and this broke some builds ❤️

felixfbecker commented:
0.14.0 is not on npm yet - could you `npm publish` it so I can remove the direct git dependency?

felixfbecker commented:
I also went through the changes on master and didn't notice any regressions.

bhs commented:
@felixfbecker 

> Please next time don't force push the branch

Yes, I did a rebase against master for reasons cited above. No way around a force-push post-rebase.

I did run the (npm) tests before pushing anything, FYI – anyway, I will `npm publish` now, I wanted to wait overnight to make sure nothing exploded (since the merge and post-merge process was not smooth sailing).

_time passes..._

https://www.npmjs.com/package/opentracing

felixfbecker commented:
Well, a `git merge master` would have achieved the same without a force push, but it's all fixed now :)

bhs commented:
/me has a lot of muscle memory built up around `rebase`, sorry :-/
